actor/todo-sql/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+        )
+        .await?;
+    let rows: Vec<Todo> = decode(&resp.rows)?;
+    Ok(rows)
+}
+
+async fn get_todo(ctx: &Context, client: &Db, url: &str) -> Result<Todo, SqlDbError> {
+    info!("Getting a todo...");
+    check_safety("url", url)?;
+    let resp = client
+        .fetch(
+            &ctx,
+            &format!("select url, title, completed, order from {} where url='{}'", TABLE_NAME, url)
+        )
+        .await?;
+    if resp.num_rows == 0 {
+        return Err(SqlDbError::new("notFound", "url not found".to_string()));
+    }
+    let mut rows: Vec<Todo> = decode(&resp.rows)?;
+    let todo = rows.remove(0);
+    Ok(todo)
+}
+
+async fn delete_all_todos(ctx: &Context, client: &Db) -> Result<(), SqlDbError> {
+    info!("Deleting all todos...");
+
+    let _resp = client
+        .execute(&ctx, &format!("delete from {}", TABLE_NAME))
+        .await?;
+
+    Ok(())
+}
+
+async fn delete_todo(ctx: &Context, client: &Db, url: &str) -> Result<(), SqlDbError> {
+    info!("Deleting a todo...");
+
+    check_safety("url", url)?;
+    let _resp = client
+        .execute(&ctx, &format!("delete from {} where url='{}'", TABLE_NAME, url))
+        .await?;
+
+    Ok(())
+}
+
+async fn handle_request(ctx: &Context, req: &HttpRequest) -> RpcResult<HttpResponse> {
+    debug!("incoming req: {:?}", req);
+
+    let trimmed_path = req.path.trim_end_matches('/');
+    let db = SqlDbSender::new();
+
+    match (req.method.as_ref(), trimmed_path) {
+        ("GET", "/") => Ok(HttpResponse {
+            body: "todo server lives at /api.  Initialize empty table with /init".to_string().into_bytes(),
+            ..Default::default()
+        }),
+        ("GET", "/init") => match create_table(ctx, &db).await {
+            Ok(()) => Ok(HttpResponse {
+                body: "table created".to_string().into_bytes(),
+                ..Default::default()
+            }),
+            Err(e) => Err(RpcError::ActorHandler(format!("creating table: {}", e.to_string()))),
+        },
+
+        ("POST", "/api") => match serde_json::from_slice(&req.body) {
+            Ok(input) => match create_todo(ctx, &db, input).await {
+                Ok(todo) => HttpResponse::json(todo, 200),
+                Err(e) => Err(RpcError::ActorHandler(format!("creating todo: {:?}", e))),
+            },
+            Err(e) => Ok(HttpResponse::bad_request(format!(
+                "malformed body: {:?}",
+                e
+            ))),
+        },
+
+        ("GET", "/api") => match get_all_todos(ctx, &db).await {
+            Ok(todos) => HttpResponse::json(todos, 200),
+            Err(e) => Err(RpcError::ActorHandler(format!("getting all todos: {}", e.to_string()))),
+        },
+
+        ("GET", url) => match get_todo(ctx,  &db, url).await {
+            Ok(todo) => HttpResponse::json(todo, 200),
+            Err(_) => Ok(HttpResponse::not_found()),
+        },
+
+        ("PATCH", url) => match serde_json::from_slice(&req.body) {
```

Comment:
While I don't personally have a problem with using the `PATCH` method, I've seen a number of case studies published where people complained about its use and instead commented an idempotent `PUT` for updates. Something about lack of support for the `PATCH` method in places but I don't know if that assertion still holds true in 2021.

Reply:
I left the REST api exactly as-is (except for adding /init). There are standard javascript test suites that validate the rest todo api and I didn't want to change anything there. I'll see if I can get more info on what the tests do

Reply:
confirmed. PATCH is part of the test suite https://github.com/TodoBackend/todo-backend-js-spec/blob/9fdf6f9390d743116f232f08efc70e118d88c7a6/js/specs.js#L22